datalad/local/copy_file.py

```diff
@@ -0,0 +1,666 @@
+            doc="""copy all source files into this DIRECTORY. This value
+            is overridden by any explicit destination path provided via
+            [CMD: --specs-from CMD][PY: 'specs_from' PY].""",
+            constraints=EnsureStr() | EnsureNone()),
+        specs_from=Parameter(
+            args=('--specs-from',),
+            metavar='SOURCE',
+            doc="""read list of source (and destination) path names from a given
+            file, or stdin (with '-'). Each line defines either a source
+            path, or a source/destination path pair (separated by a null byte
+            character).[PY:  Alternatively, a list of 2-tuples with
+            source/destination pairs can be given. PY]"""),
+        message=save_message_opt,
+    )
+
+    _examples_ = [
+        dict(
+            text="Copy a file into a dataset 'myds' using a path and a target "
+                 "directory specification, and save its addition to 'myds'",
+            code_py="""\
+            copy_file('path/to/myfile', target_dir='path/to/myds/',
+                      dataset='path/to/myds')""",
+            code_cmd="""\
+            datalad copy-file path/to/myfile -t path/to/myds/ -d path/to/myds"""),
+        dict(
+            text="Copy a file to a dataset 'myds' and save it under a new name "
+                 "by providing two paths",
+            code_py="""\
+            copy_file(path=['path/to/myfile', 'path/to/myds/newname'],
+                      dataset='path/to/myds')""",
+            code_cmd="""\
+            datalad copy-file path/to/myfile path/to/myds/new -d path/to/myds"""),
+        dict(
+            text="Copy a file into a dataset without saving it",
+            code_py="copy_file('path/to/myfile', target_dir='path/to/myds/')",
+            code_cmd="datalad copy-file path/to/myfile -t path/to/myds"),
+        dict(
+            text="Copy a directory and its subdirectories into a dataset 'myds'"
+                 " and save the addition in 'myds'",
+            code_py="""\
+            copy_file('path/to/dir/', target_dir='path/to/myds/',
+                      recursive=True, dataset='path/to/myds')""",
+            code_cmd="""\
+            datalad copy-file path/to/dir -t path/to/myds -r -d path/to/myds"""),
+        dict(
+            text="Copy files using a path and optionally target specification "
+                 "from a file",
+            code_py="""\
+            copy_file(dataset='path/to/myds', target_dir='path/to/myds',
+                      specs_from='path/to/specfile')""",
+            code_cmd="""\
+            datalad copy-file -d path/to/myds -t path/to/myds --specs-from specfile"""
+        ),
+        dict(
+            text="Read a specification from stdin: Pipe the output of a find "
+                 "command into the copy-file command",
+            code_py="",
+            code_cmd="""\
+            find <expr> | datalad copy-file -d myds -t path/to/myds --specs-from -"""
```

Comment:
Hmm, shouldn't the dataset path match between `-d` and `-t`?

Reply:
oh true, thanks for catching that!